docs/Branches.md

```diff
@@ -0,0 +1,73 @@
+# What Builds With What?
+
+## The Development Branches
+
+| Swift  | LLVM*  | LLDB 
+| ------ | ------ | ------
+| master | stable | stable
+
+`master` is the place for active development on Swift. If you're just working on Swift, that's where you'll spend most of your time.
+
+LLVM repos automatically merge changes from the latest release branch (see below) into `stable`. This generally means `stable` is just an alias for the latest release branch. If you want to do Swift-related development on LLVM projects, see "The Upstream Branches" below.
+
+LLDB is a bit more complicated. Because it's an [LLVM project][lldb] originally, it follows the LLVM branch names. However, it also contains extra support for Swift, and so it depends on being in sync with Swift as well. Work on LLDB is usually paired with work on Swift, so following Swift's branch is the way to go.
+
+\* "LLVM" refers to several repositories from the LLVM project that do *not* depend on Swift: LLVM, Clang, and compiler-rt. LLDB is different because it imports some of Swift's own headers and thus depends on Swift.
+
+  [release manager]: https://swift.org/blog/swift-4-0-release-process/
+  [lldb]: http://lldb.llvm.org
+
+## The Release Branches
+
+| Swift            | LLVM*            | LLDB 
+| ---------------- | ---------------- | ----------------
+| swift-x.y-branch | swift-x.y-branch | swift-x.y-branch
+
+At some point before a release, a *release branch* will be created in every repository with a name like `swift-4.0-branch`. (The actual number is chosen by Apple.) After the branch has been created, commits must make it to this branch to make it into the release. In some cases, the [release manager][] for the branch will decide to merge in all additional changes from `master`; otherwise, cherry-picking changes and making a new pull request is the way to go. If there are any "patch" releases (e.g. Swift 4.0.1), they will also come from this branch.
+
+For some releases, the release branch for the LLVM projects (LLVM, Clang, compiler-rt, and LLDB) will be based on the corresponding llvm.org release branch.
+
+
+## The Upstream Branches
+
+| Swift       | LLVM                | LLDB 
+| ----------- | ------------------- | -------------------
+| master-next | upstream-with-swift | upstream-with-swift
+
+`upstream-with-swift` is a branch for LLVM that includes all changes necessary to support Swift. Changes from llvm.org's master branch are automatically merged in. Why isn't this just `stable`? Well, because LLVM changes *very* rapidly, and that wouldn't be very stable. However, release branches for LLVM and Clang *do* include the latest upstream changes, and so they come from this branch. That means they need to include any necessary Swift support.
```

Comment:
Hmm... I re-read the document and in the context of everything, it is clear what you are saying.

On the rebranching master-next merge, I think it is important to discuss since all of the branches are related at that point in time by a coordinated merge. The fact that coordinating merges exist is important. Also we should discuss how at that time we perform the "-s ours" merge.